okio/jvm/src/main/java/okio/SegmentedByteString.kt

```diff
@@ -222,40 +226,65 @@ internal class SegmentedByteString(buffer: Buffer, byteCount: Int) : ByteString(
     otherOffset: Int,
     byteCount: Int
   ): Boolean {
-    var offset = offset
-    var otherOffset = otherOffset
-    var byteCount = byteCount
     if (offset < 0 || offset > size - byteCount ||
         otherOffset < 0 || otherOffset > other.size - byteCount) {
       return false
     }
     // Go segment-by-segment through this, comparing ranges of arrays.
-    var s = segment(offset)
-    while (byteCount > 0) {
-      val segmentOffset = if (s == 0) 0 else directory[s - 1]
-      val segmentSize = directory[s] - segmentOffset
-      val stepSize = minOf(byteCount, segmentOffset + segmentSize - offset)
-      val segmentPos = directory[segments.size + s]
-      val arrayOffset = offset - segmentOffset + segmentPos
-      if (!arrayRangeEquals(segments[s], arrayOffset, other, otherOffset, stepSize)) return false
-      offset += stepSize
-      otherOffset += stepSize
-      byteCount -= stepSize
-      s++
+    var otherOffset = otherOffset
+    processSegments(offset, offset + byteCount) { data, offset, byteCount ->
+      if (!arrayRangeEquals(data, offset, other, otherOffset, byteCount)) return false
+      otherOffset += byteCount
     }
     return true
   }
 
-  override fun indexOf(other: ByteArray, fromIndex: Int) = toByteString().indexOf(other, fromIndex)
-
-  override fun lastIndexOf(other: ByteArray, fromIndex: Int) = toByteString().lastIndexOf(other,
-      fromIndex)
-
   /** Returns a copy as a non-segmented byte string.  */
   private fun toByteString() = ByteString(toByteArray())
 
   override fun internalArray() = toByteArray()
 
+  /** Processes all segments, invoking `yield` with the ByteArray and range of valid data. */
+  private inline fun processSegments(
+    yield: (data: ByteArray, offset: Int, byteCount: Int) -> Unit
```

Comment:
I would prefer these not be called "yield" since Kotlin uses that for suspending callbacks: http://kotlinlang.org/api/latest/jvm/stdlib/kotlin.coroutines.experimental/-sequence-builder/yield.html. This helper is kind of like a fancy `forEach` which uses "action": http://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/for-each.html

Reply:
I can change it to `action`. Maybe also change the name to `forEachSegment`?

I will also mention that I used `yield` for the UTF-8 decoding methods in the common module. I could change those to `action` as well but it feels a little out of scope. Maybe a separate pull request?

Reply:
I'd be happy with `forEachSegment`. I really like how much this cleans up the other code!

Reply:
Thumbs up on `forEachSegment`.

Reply:
Its functions like this that make me want to convert **everything** to Kotlin